### PR TITLE
Serializable messages

### PR DESCRIPTION
**Pull request by mattprecious**
@danrice-square @swankjesse @JakeWharton

### CODE REVIEW

File: wire-runtime/src/main/java/com/squareup/wire/MessageSerializedForm.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ * limitations under the License.
+ */
+package com.squareup.wire;
+
+import java.io.IOException;
+import java.io.ObjectStreamException;
+import java.io.Serializable;
+import java.io.StreamCorruptedException;
+
+public final class MessageSerializedForm implements Serializable {
```
Comment (JakeWharton) [inline wire-runtime/src/main/java/com/squareup/wire/MessageSerializedForm.java]: Not public?

File: wire-runtime/src/main/java/com/squareup/wire/MessageSerializedForm.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  private final Class<? extends Message> messageClass;
+
+  public MessageSerializedForm(Message message, Class<? extends Message> messageClass) {
+    this.bytes = message.toByteArray();
+    this.messageClass = messageClass;
+  }
+
+  Object readResolve() throws ObjectStreamException {
+    try {
+      return new Wire().parseFrom(bytes, messageClass);
```
Comment (danrice-square) [inline wire-runtime/src/main/java/com/squareup/wire/MessageSerializedForm.java]: This implies that the serialized classes can't contain extensions, i.e., any extensions will end up in the UnknownField bin. Could we serialize a list of extension names as needed?

Comment (mattprecious) [inline wire-runtime/src/main/java/com/squareup/wire/MessageSerializedForm.java]: I don't see any clean way of doing this from my quick look around at how extensions are done. The extension classes are lost once they're added to the registry in the Wire constructor... can this be a follow-up or just omitted completely?

**Comment by mattprecious** [comment]
Needed to emit `serialVersionUID` in all the messages :(

File: wire-runtime/src/main/java/com/squareup/wire/MessageSerializedForm.java (hunk 2 above)
Comment (danrice-square) [inline wire-runtime/src/main/java/com/squareup/wire/MessageSerializedForm.java]: Seems like something we could follow up with, provided we allow for there to be additional fields in `MessageSerializedForm` in the future.

**Comment by danrice-square** [comment]
Do we actually need to generate code for all of descriptor.proto?

File: wire-runtime/src/test/java/com/google/protobuf/MessageOptions.java [hunk 3]
```diff
@@ -51,7 +52,7 @@
   /**
    * The parser stores options it doesn't recognize here. See above.
    */
-  @ProtoField(tag = 999, label = REPEATED)
+  @ProtoField(tag = 999, label = REPEATED, messageType = UninterpretedOption.class)
```
Comment (danrice-square) [inline wire-runtime/src/test/java/com/google/protobuf/MessageOptions.java]: Thanks for fixing this.

**Comment by mattprecious** [comment]
No. They're generated with the `gen_tests` script, though, so it gets annoying dealing with git changes. It would be great if they were omitted from the script.

File: wire-runtime/src/main/java/com/squareup/wire/MessageSerializedForm.java (hunk 2 above)
Comment (mattprecious) [inline wire-runtime/src/main/java/com/squareup/wire/MessageSerializedForm.java]: Was there something in particular you had in mind? I think fields can be added here in the future without any incompatibility with the current structure. When an old version gets deserialized the new fields will be null, and then we can treat it like a message without extensions.

If anything more complicated arises we can branch based on the version UID.

**Comment by swankjesse** [comment]
LGTM

**Comment by danrice-square** [comment]
LGTM

**Pull request merged**